moz_minions/boss.py

```diff
@@ -1,58 +1,127 @@
+from watchdog.observers import Observer
+from watchdog.events import PatternMatchingEventHandler
 from minions import ShellMinion
 
 
+class JsonHandler(PatternMatchingEventHandler):
+    def set_handler(self, oncreated=None, onmodified=None, ondeleted=None):
+        self.create_handler = oncreated
+        self.modify_handler = onmodified
+        self.delete_handler = ondeleted
+
+    def on_created(self, event):
+        self.create_handler(event.src_path)
+
+    def on_modified(self, event):
+        self.modify_handler(event.src_path)
+
+    def on_deleted(self, event):
+        self.delete_handler(event.src_path)
+
+
 class Boss(object):
-    default_path = None
-    workers = []
+    workers = {}
+    dirpath = '.'
+    output = None
 
-    def __init__(self):
+    def __init__(self, dirpath='.', output='output'):
         '''
         local path for load config
         '''
+        if os.path.isdir(dirpath):
+            self.dirpath = dirpath
+        else:
+            print(dirpath + " is invalid, use default path instead")
+        self.output = output
+        if not os.path.isdir(output):
+            print("target directory " + output + " doesn't exist, creating..")
+            os.makedirs(output)
+
         self.scheduler = BackgroundScheduler()
         self.scheduler.start()
-        if self.default_path:
-            self.load(self.default_path)
+
+        self.load_dir(dirpath)
+
+        event_handler = JsonHandler(patterns=["*.json"],
+                                    ignore_directories=True)
+        event_handler.set_handler(oncreated=self.load,
+                                  onmodified=self.load,
+                                  ondeleted=self.remove)
+        observer = Observer()
+        observer.schedule(event_handler, self.dirpath, recursive=True)
+        observer.start()
 
     def load_dir(self, folder):
         (dirpath, dirnames, filenames) = os.walk(folder).next()
-        for fn in filenames:
-            self.load(os.path.join(dirpath, fn))
+        for fname in filenames:
+            if 'json' in fname[-4:]:
+                self.load(os.path.join(dirpath, fname))
 
     def load(self, fp):
         '''
-        given a file
-        TBI: directory
+        given a json file, load and create a task run regularly
         '''
+        # for debugging
+        print(fp + " was loaded!")
         with open(fp) as in_data:
-            data = json.load(in_data)
-            minion = ShellMinion(**data)
-            self.workers.append(minion)
-            self.scheduler.add_job(minion.collect, 'interval',
-                                   name=minion.name+'_'+minion.serial,
-                                   seconds=30
-                                   )
+            try:
+                data = json.load(in_data)
+            except ValueError as e:
+                print(fp + " loaded failed: " + e.message)
+                return None
```

Comment:
Just a quick question. After "return None", in load_dir(dirpath) funtion, what will happen if there is nothing loaded?